Added kwarg handling to rect.c

jonotassia opened the pull request:
Implemented the below changes to display.c functions (per #3852) to add kwarg support:

- Kwarg support in display.c
- Doc updates in display.rst
- Test cases updated and passed

Functions covered:

- `pg_rect_clipline` 
- `pg_rect_scale_by`
- `pg_rect_scale_by_ip`
- `pg_rect_unionall`
- `pg_rect_unionall_ip`
- `pg_rect_collidelist`
- `pg_rect_collidelistall`
- `pg_rect_collidedict`
- `pg_rect_collidedictall`

jonotassia commented:
I added and ran tests for all of these, and they are working across the board, but I'm not sure how best to handle keyword arguments when the function is overloaded in the functions below:

- `pg_rect_clipline` 
- `pg_rect_scale_by`
- `pg_rect_scale_by_ip`

In these cases, arguments could be passed in a number of ways:

`clipline`
```
    @overload
    def clipline(
        self, x1: float, x2: float, x3: float, x4: float
    ) -> Union[Tuple[Tuple[int, int], Tuple[int, int]], Tuple[()]]: ...
    @overload
    def clipline(
        self, first_coordinate: Coordinate, second_coordinate: Coordinate
    ) -> Union[Tuple[Tuple[int, int], Tuple[int, int]], Tuple[()]]: ...
    @overload
    def clipline(
        self, rect_arg: RectValue
    ) -> Union[Tuple[Tuple[int, int], Tuple[int, int]], Tuple[()]]: ...
```

`scale_by` and `scale_by_ip`
```
    @overload
    def scale_by(self, x: float, y: float) -> Rect: ...
    @overload
    def scale_by(self, scale_by: Coordinate) -> Rect: ...
    @overload
    def scale_by_ip(self, x: float, y: float) -> None: ...
    @overload
    def scale_by_ip(self, scale_by: Coordinate) -> None: ...
```

I'd imagine most users would expect to be able to pass the arguments in the overloaded functions as well, ie `scale_by(scale_by=(x, y))` as well as `scale_by(x=x, y=y)`, but I'm not sure how best to handle the different ways to interface with the function without adding in a number of other variables and explicitly handling them, which seems a little messy. 

Do you have any recommendations on how to handle kwargs with overloaded functions?

illume commented on src_c/rect.c:
```diff
@@ -565,14 +570,16 @@ pg_rect_unionall(pgRectObject *self, PyObject *args)
     int t, l, b, r;
 
-    if (!PyArg_ParseTuple(args, "O", &list)) {
+    static char *keywords[] = {"rect_sequence", NULL};
```
Probably this could be named “rects”?

illume commented on src_c/rect.c:
```diff
@@ -721,15 +728,17 @@ pg_rect_colliderect(pgRectObject *self, PyObject *const *args,
     PyObject *ret = NULL;
 
-    if (!PyArg_ParseTuple(args, "O", &list)) {
+    static char *keywords[] = {"rect_list", NULL};
```
“rects”?

illume commented on test/rect_test.py:
```diff
@@ -2555,6 +2785,29 @@ def test_collidelistall_list_of_object_with_rect_property(self):
         self.assertFalse(r.collidelistall(f))
 
+    def test_collidelistall__kwargs(self):
+        # __doc__ (as of 2008-08-02) for pygame.rect.Rect.collidelistall:
```
These __doc__ comments could be removed in these new tests.

illume commented:
Very good. I left a few notes for your consideration.

As to your question…

> I'd imagine most users would expect to be able to pass the arguments in the overloaded functions as well, ie scale_by(scale_by=(x, y)) as well as scale_by(x=x, y=y), but I'm not sure how best to handle the different ways to interface with the function without adding in a number of other variables and explicitly handling them, which seems a little messy.

> Do you have any recommendations on how to handle kwargs with overloaded functions?

Yeah, people will probably want to call them like that. Good point.

You add all of the keyword arguments. Then if they are not used they are set to NULL if not used. So you can check the user has done something sensible. Raise a ValueError if an unsupported combination is used. Like x and scale_by but not y.

jonotassia commented:
Alright, this should all be in working order now (though I'll keep an eye on the automated checks in case there are any compilation errors). 

Updates:

1. I've now added handling for all the different overloaded versions of `clipline`, `scale_by`, and `scale_by_ip`. It added more lines of code than I initially expected to make sure the keywords were being handled before hitting `PyArg_ParseTupleAndKeywords`, so open to any suggestions if you can see a better way to account for the different keyword options.

2. I've also added test cases to ensure exceptions are appropriately raised if a user passes incorrect inputs to the functions via kwargs.

3. Lastly, I've updated the stub files per your previous suggestions for `collidelist` and `unionlistall` so that the keywords for both of those are simply `rects` rather than `rect_list` and `rect_sequence`.

jonotassia commented:
Looks like Python versions over 3.7 are failing the below test cases across the board. I'll take a look next week and try to get that fixed.

```
  ======================================================================
  ERROR: test_scale_by_ip__kwargs (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 1072, in test_scale_by_ip__kwargs
      r2.scale_by_ip(x=2, y=4)
  SystemError: Function returned a NULL result without setting an exception
  
  ======================================================================
  ERROR: test_scale_by_ip__larger (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 1036, in test_scale_by_ip__larger
      r2.scale_by_ip(2)
  SystemError: Function returned a NULL result without setting an exception
  
  ======================================================================
  ERROR: test_scale_by_ip__smaller (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 1050, in test_scale_by_ip__smaller
      r2.scale_by_ip(0.5)
  SystemError: Function returned a NULL result without setting an exception
  
  ======================================================================
  ERROR: test_scale_by_ip__subzero (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 1063, in test_scale_by_ip__subzero
      r.scale_by_ip(0)
  SystemError: Function returned a NULL result without setting an exception
  
  ======================================================================
  FAIL: test_scale_by__kwarg_exceptions (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle using
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 867, in test_scale_by__kwarg_exceptions
      r.scale_by(scale_by=(1, 2), y=1)
  AssertionError: SystemError not raised
  
  ======================================================================
  FAIL: test_scale_by__larger (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 890, in test_scale_by__larger
      self.assertEqual(r.left - 3, r2.left)
  AssertionError: -1 != 2
  
  ======================================================================
  FAIL: test_scale_by__larger_kwargs (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 926, in test_scale_by__larger_kwargs
      self.assertEqual(r.left - 3, r2.left)
  AssertionError: -1 != 2
  
  ======================================================================
  FAIL: test_scale_by__larger_kwargs_scale_by (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 908, in test_scale_by__larger_kwargs_scale_by
      self.assertEqual(r.left - 3, r2.left)
  AssertionError: -1 != 2
  
  ======================================================================
  FAIL: test_scale_by__larger_single_argument (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 837, in test_scale_by__larger_single_argument
      self.assertEqual(r.left - 3, r2.left)
  AssertionError: -1 != 2
  
  ======================================================================
  FAIL: test_scale_by__larger_single_argument_kwarg (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle using
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 851, in test_scale_by__larger_single_argument_kwarg
      self.assertEqual(r.left - 3, r2.left)
  AssertionError: -1 != 2
  
  ======================================================================
  FAIL: test_scale_by__smaller (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 941, in test_scale_by__smaller
      self.assertEqual(r.left + 2, r2.left)
  AssertionError: 4 != 2
  
  ======================================================================
  FAIL: test_scale_by__smaller_single_argument (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 875, in test_scale_by__smaller_single_argument
      self.assertEqual(r.left + 2, r2.left)
  AssertionError: 4 != 2
  
  ======================================================================
  FAIL: test_scale_by__subzero (pygame.tests.rect_test.RectTypeTest)
  The scale method scales around the center of the rectangle
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/tmp/tmp.bjwhdYnITL/venv/site-packages/pygame/tests/rect_test.py", line 957, in test_scale_by__subzero
      self.assertEqual(r.centerx - r.w * 10 / 2, rx1.x)
  AssertionError: -25.0 != 2
```

illume commented:
What do you think about removing the scaleby changes, and then the other stuff can be merged in?
I guess it's only the scaleby one that is having problems.

jonotassia commented:
I actually think I may have found the issue this morning: I accidentally initialised the temporary values as `ints` rather than `floats` in `scale_by_ip` , so that would have lead to `PyArg_ParseTupleAndKeywords` to fail and return `NULL`, and would probably also explain the large difference in return values in some of the test cases. 

I'm not sure why the handling was different Python 3.8 and higher, but we'll see if this passes (still passing in 3.7). I'll check in shortly when the automated checks finish running.

illume reviewed (approved):
Awesome. Thanks. 🎉🎈
